README.md
# OCaml-migrate-parsetree

**DEPRECATED** This package is deprecated. Please, use
[ppxlib](https://github.com/ocaml-ppx/ppxlib) instead.
See https://ocaml.org/changelog/2023-10-23-omp-deprecation.

Convert OCaml parsetrees between different major versions

This library converts between parsetrees of different OCaml versions.

Supported versions are 4.02, 4.03, 4.04, 4.05, 4.06, 4.07, 4.08, 4.09,
4.10, 4.11, 4.12 and 4.13. For each version, there is a snapshot of the
parsetree and conversion functions to the next and/or previous version.

## Asts

```ocaml
module Ast_{402,403,404,405,406,407,408,409,410,411,412,413} : sig

  (* Version specific copy of AST *)
  module Asttypes
  module Parsetree

  (* Magic numbers used for marshalling *)
  module Config : sig
    val ast_impl_magic_number : string
    val ast_intf_magic_number : string
  end
end
```

These embed copies of AST definitions for each supported OCaml major version.

The AST matching the version of the OCaml toolchain will contain
equalities relating the copy of types to the definitions from
compiler-libs.  For instance, when installed with OCaml 4.04.x the
type `Ast_404.Parsetree.expression` is equal to the type
`Parsetree.expression` from the compiler libraries.

## Migration modules

For each pair of versions `$(n)` and `$(n+1)`, the two modules
`Migrate_parsetree_$(n)_$(n+1)` and `Migrate_parsetree_$(n+1)_$(n)`
convert the AST forward and backward.

The forward conversion is total while the backward conversion is
partial: when a feature is not available in a previous version of the
parsetree, a `Migrate_parsetree_def.Migration_error` exception is
raised detailing the failure case.

## Adding a new OCaml version

We use [Cinaps](https://github.com/janestreet/cinaps) to generate
boilerplate.  You can install it via opam: `opam install cinaps`.

Add the new version in
[src/cinaps_helpers/cinaps_helpers.ml](https://github.com/ocaml-ppx/ocaml-migrate-parsetree/blob/master/src/cinaps_helpers/cinaps_helpers.ml)
`supported_versions`.

Copy the last `src/ast_xxx.ml` file to `src/ast_<new_version>.ml`,
then go over the file and update each sub-module by replacing its
signature and implementation with the code from the compiler. For the
`Config` sub-module, update the two variables with the values in
`utils/config.mlp` in the compiler source tree.

Once this is done, call:

    $ dune exec tools/add_special_comments.exe src/ast_<new_version>.ml

Then diff the `src/ast_xxx.ml` and `src/ast_<new_version>.ml` and go
over the diff to make sure the difference are relevant. The `ast_...`
files require some adjustments which should pop up when you do this
diff. Port the old adjustments to the new file as required.

Add migration functions. In the commands below, set $OLD and $NEW to the
appropriate version numbers, e.g. 408 and 409:
- Manually compile the asts (`ocamlc -c src/ast_{$NEW,$OLD}.ml -I +compiler-libs -I _build/default/src/.migrate_parsetree.objs/byte/`)
- Using `tools/gencopy.exe` (`dune build tools/gencopy.exe`), generate copy code to and from previous version:
```
_build/default/tools/gencopy.exe -I . -I src/ -I +compiler-libs -map Ast_${NEW}:Ast_${OLD} Ast_${NEW}.Parsetree.{expression,pattern,core_type,toplevel_phrase} > src/migrate_${NEW}_${OLD}.ml
_build/default/tools/gencopy.exe -I . -I src/ -I +compiler-libs -map Ast_${OLD}:Ast_${NEW} Ast_${OLD}.Parsetree.{expression,pattern,core_type,toplevel_phrase} > src/migrate_${OLD}_${NEW}.ml
```
- Fix the generated code by implementing new cases

*TODO*: specialize and improve gencopy for these cases

At any time, you can expand boilerplate code by running `make cinaps`.

Update build system:
- make sure `make cinaps` reaches a fixed point :)
- `make` should succeed


CHANGES.md
unreleased
------------------------

- Deprecate the project in favor of ppxlib (#126, @pitag-ha)


2.4.0 2022-06-17 Valencia
--------------------------

- Add support for 5.0 (#122, @pitag-ha)

v2.3.0 2021-12-07 Valencia
--------------------------

- Add support for 4.14 (#119, @kit-ty-kate)

v2.2.0 2021-06-21 Paris
--------------------------

- Add support for 4.13 (#114, @kit-ty-kate)

v2.1.0 2020-10-22 Bordeaux
--------------------------

- Add support for 4.12 (#107, @ceastlund)

v2.0.0 2020-08-11 Canterbury
----------------------------

- No longer expose the unwrapped modules (#94, @jonludlam)

- Remove everything but Ast versions and upgrade/downgrade
  conversions (#103, @ceastlund)

v1.7.3 2020-05-07 Canterbury
----------------------------

- Fix magic numbers for the 4.11 ast (#96, @hhugo)

v1.7.2 2020-04-20 Canterbury
----------------------------

- Remove toplevel `Option` module accidentally added in 1.7.0

v1.7.1 2020-04-15 Canterbury
----------------------------

- Fix build with OCaml < 4.08

v1.7.0 2020-04-09 Canterbury
----------------------------

- Add support for 4.11 (#92, @diml)

v1.6.0 2020-02-10 Moscow
------------------------

- Preserve compiler version of binary ASTs across transformation (#79,
  @aantron)

- Allow not exiting on error (#83, @aantron)

v1.5.0 2019-11-18
-----------------

- Add support for 4.10 (#86, @diml)

- Infer file kind (interface or implementation) for binary ASTs that
  have no extension (#80, @aantron)

- Add ?argv argument to Driver.run_main (#82, @aantron)

v1.4.0 2019-07-04 London
------------------------

- Initial support for 4.09, tested with 4.09+beta1 (#76, @hhugo)

- When encoding errors into the AST, duplicate the error message for
  "ocaml.error" nodes for OCaml versions < 4.08 (#75, @xclerc)

v1.3.1 2019-05-20 London
------------------------

- Make sure opening `Ast_408` doesn't shadow `Int` or `Misc` (#71,
  @hhugo)

- Fix a couple of issues related to upgrading the AST from 4.07 to
  4.08 (#71, @hhugo)

v1.3.0 2019-05-08 London
------------------------

- Get rid of the ocamlbuild plugin. Nobody is using it in opam and it
  is more work to maintain (#63, @diml)

- Set `Location.input_name` to the original filename when reading a
  binary AST (#66, @diml)

- Add support 4.08 (#70, @xclerc)

v1.2.0 2018-12-19 London
------------------------

- Remove unused ocamlfind dependency in the opam file (#53, @diml)

- Add `--print-transformations` to list registered transformations
  (#55, @rgrinberg)

- Fix Windows compatibility by setting the output to binary mode when
  writing a binary ast (#57, #59, @bryphe and @dra27)

- Switch to dune and opam 2.0 (#58, #60, @diml)

v1.1.0 2018-09-05 London
------------------------

- Allow ppx rewriters to specify when they should be applied

v1.0.11 2018-06-06 London
-------------------------

- Fix handling of `--impl/--intf`. Before the driver would crash if
  the file extension was neither `.ml` nor `.mli`

v1.0.10 2018-04-19 London
-------------------------

- Add support for OCaml 4.07

v1.0.9 2018-03-20 New York
--------------------------

- Fix an issue where cookies set from the command line sometimes
  disappeared

v1.0.8 2018-03-15 London
------------------------

- Add a `--null` argument to suppress the output. This is used to
  write linters
- Use the new generic ppx driver support of jbuilder

v1.0.7 2017-10-31 Paris
-----------------------

Contributed by @hhugo:
- update Magic Number for 4.06
- fix some compilation warnings

v1.0.6 2017-10-11 Paris
-----------------------

Fix generation of `Migrate_parsetree` module.

v1.0.5 2017-10-02 Paris
-----------------------

Resynchronize with trunk.
Add a migrating version of Parse module.

v1.0.4 2017-08-22 Paris
-----------------------

Resynchronize with trunk. Contributed by Xavier Clerc, @xclerc.

v1.0.3 2017-08-11 Paris
-----------------------

Add a shallow identity mapper (suggested by Anton Bachin, @aantron).

v1.0.2 2017-07-28 Paris
-----------------------

Synchronize with 4.06 AST with trunk.
Accept --cookie arguments also when run in --as-ppx mode.

v1.0.1 2017-06-06 Paris
-----------------------

Add support for trunk version (as of today...).

v1.0 2017-04-17 Paris
---------------------

Driver: add --as-pp and --embed-errors flags.

    --embed-errors causes the driver to embed exceptions raised by
    rewriters as extension points in the Ast

    --as-pp is a shorthand for: --dump-ast --embed-errors

Expose more primitives for embedding the driver.

Fix bug where `reset_args` functions where not being called.
Fix "OCaml OCaml" in error messages (contributed by Adrien Guatto).

v0.7 2017-03-21 Mâcon
---------------------

Fix findlib predicates:
- replace `omp_driver` by `ppx_driver`
- replace `-custom_ppx` by `-custom_ppx,-ppx_driver`

v0.6 2017-03-21 Mâcon
---------------------

Add documentation, examples, etc.

v0.5 2017-03-11 Mâcon
---------------------

Specify ocamlfind dependency in opam file (@yunxing).

v0.4 2017-03-10 Mâcon
---------------------

API cleanup and extension. Added driver. Switch to jbuilder.

v0.3 2017-02-16 Paris
----------------------

Use `-no-alias-deps` to prevent linking failure of `Compiler_libs` (referencing `Parsetree` and `Asttypes` which have no implementation).

v0.2 2017-02-07 London
----------------------

Install CMXS too (contributed @vbmithr).

v0.1 2017-02-02 London
----------------------

First release.


src/cinaps.ml
open StdLabels
open Printf

let nl () = printf "\n"

let supported_versions = [
  ("402", "4.02");
  ("403", "4.03");
  ("404", "4.04");
  ("405", "4.05");
]

let qualified_types = [
  "Parsetree",
  [ "structure"
  ; "signature"
  ; "toplevel_phrase"
  ; "core_type"
  ; "expression"
  ; "pattern"
  ; "case"
  ; "type_declaration"
  ; "type_extension"
  ; "extension_constructor"
  ];

  "Outcometree",
  [ "out_value"
  ; "out_type"
  ; "out_class_type"
  ; "out_module_type"
  ; "out_sig_item"
  ; "out_type_extension"
  ; "out_phrase"
  ];

  "Ast_mapper",
  [ "mapper"
  ];
]

let all_types = List.concat (List.map ~f:snd qualified_types)

let foreach_module f =
  nl ();
  List.iter qualified_types ~f:(fun (m, types) -> f m types)

let foreach_type f =
  nl ();
  foreach_module (fun m -> List.iter ~f:(f m))

let foreach_version f =
  nl ();
  List.iter supported_versions ~f:(fun (suffix, version) -> f suffix version)

let foreach_version_pair f =
  nl ();
  let rec aux = function
    | (x,_) :: ((y,_) :: _ as tail) -> f x y; aux tail
    | [_] | [] -> ()
  in
  aux supported_versions

let with_then_and () =
  let first = ref true in fun oc ->
    output_string oc (if !first then "with" else "and");
    first := false
